llvm/lib/SYCLLowerIR/LowerESIMD.cpp

```diff
@@ -0,0 +1,1316 @@
+      GenXIntrinsic::getGenXIntrinsicPrefix() + Desc.GenXSpelling + Suffix);
+
+  SmallVector<Value *, 16> GenXArgs;
+  createESIMDIntrinsicArgs(Desc, GenXArgs, CI, FE);
+
+  SmallVector<Type *, 16> GenXOverloadedTypes;
+  if (GenXIntrinsic::isOverloadedRet(ID))
+    GenXOverloadedTypes.push_back(CI.getType());
+  for (unsigned i = 0; i < GenXArgs.size(); ++i)
+    if (GenXIntrinsic::isOverloadedArg(ID, i))
+      GenXOverloadedTypes.push_back(GenXArgs[i]->getType());
+
+  Function *NewFDecl = GenXIntrinsic::getGenXDeclaration(CI.getModule(), ID,
+                                                         GenXOverloadedTypes);
+
+  Instruction *NewCI = IntrinsicInst::Create(
+      NewFDecl, GenXArgs,
+      NewFDecl->getReturnType()->isVoidTy() ? "" : CI.getName() + ".esimd",
+      &CI);
+  NewCI = addCastInstIfNeeded(&CI, NewCI);
+  CI.replaceAllUsesWith(NewCI);
+  CI.eraseFromParent();
+}
+
+static std::string getMDString(MDNode *N, unsigned I) {
+  if (!N)
+    return "";
+
+  Metadata *Op = N->getOperand(I);
+  if (!Op)
+    return "";
+
+  if (MDString *Str = dyn_cast<MDString>(Op)) {
+    return Str->getString().str();
+  }
+
+  return "";
+}
+
+void SYCLLowerESIMDLegacyPass::generateKernelMetadata(Module &M) {
+  if (M.getNamedMetadata(GENX_KERNEL_METADATA))
+    return;
+
+  auto Kernels = M.getOrInsertNamedMetadata(GENX_KERNEL_METADATA);
+  assert(Kernels->getNumOperands() == 0 && "metadata out of sync");
+
+  LLVMContext &Ctx = M.getContext();
+  Type *I32Ty = Type::getInt32Ty(Ctx);
+
+  std::string TargetTriple = M.getTargetTriple();
+  llvm::Triple T(TargetTriple);
+  T.setArchName("genx64");
+  TargetTriple = T.str();
+  M.setTargetTriple(TargetTriple);
+
+  enum { AK_NORMAL, AK_SAMPLER, AK_SURFACE, AK_VME };
+  enum { IK_NORMAL, IK_INPUT, IK_OUTPUT, IK_INPUT_OUTPUT };
+
+  for (auto &F : M.functions()) {
+    // Skip non-SIMD kernels.
+    if (F.getCallingConv() != CallingConv::SPIR_KERNEL ||
+        F.getMetadata("sycl_explicit_simd") == nullptr)
+      continue;
+
+    // Metadata node containing N i32s, where N is the number of kernel
+    // arguments, and each i32 is the kind of argument,  one of:
+    //     0 = general, 1 = sampler, 2 = surface, 3 = vme
+    // (the same values as in the "kind" field of an "input_info" record in a
+    // vISA kernel.
+    SmallVector<Metadata *, 8> ArgKinds;
+
+    // Optional, not supported for compute
+    SmallVector<Metadata *, 8> ArgInOutKinds;
+
+    // Metadata node describing N strings where N is the number of kernel
+    // arguments, each string describing argument type in OpenCL.
+    // required for running on top of OpenCL runtime.
+    SmallVector<Metadata *, 8> ArgTypeDescs;
+
+    auto *KernelArgTypes = F.getMetadata("kernel_arg_type");
+    unsigned Idx = 0;
+
+    // Iterate argument list to gather argument kinds and generate argument
+    // descriptors.
+    for (auto AI = F.arg_begin(), AE = F.arg_end(); AI != AE; ++AI) {
```

Comment:
Nice to have:  range-based for loop over F.args().

Reply:
ok